File: blaise-svg/src/main/java/com/googlecode/blaisemath/graphics/svg/SVGUtils.java
/**
 * SVGUtils.java
 * Created on May 5, 2015
 */
package com.googlecode.blaisemath.graphics.svg;

/*
 * #%L
 * BlaiseSVG
 * --
 * Copyright (C) 2014 - 2021 Elisha Peterson
 * --
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import com.googlecode.blaisemath.style.Marker;
import com.googlecode.blaisemath.svg.SVGPath;
import com.googlecode.blaisemath.util.AffineTransformBuilder;
import java.awt.geom.Path2D;
import java.awt.geom.Point2D;
import static java.util.Objects.requireNonNull;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Utility class for SVG paths.
 * @author petereb1
 */
public class SVGUtils {

    private static final Logger LOG = Logger.getLogger(SVGUtils.class.getName());

    private static final Pattern LEN_PATTERN = Pattern.compile("^auto$|^[+-]?[0-9]+\\.?([0-9]+)?(px|em|ex|%|in|cm|mm|pt|pc)?$");
    
    private SVGUtils() {
    }
    
    /**
     * Convert the given path string to a marker with given size.
     * @param svgPath SVG path string
     * @param sz size of the resulting marker, as the side length/diameter
     * @return marker
     */
    public static final Marker pathToMarker(String svgPath, final float sz) {
        final Path2D path = SVGPath.shapeConverter().convert(new SVGPath(svgPath));
        if (path == null) {
            
        }
        return (Point2D point, double orientation, float markerRadius) -> {
            double scale = markerRadius/(.5*sz);
            return path.createTransformedShape(new AffineTransformBuilder()
                    .translate(point.getX()-markerRadius, point.getY()-markerRadius)
                    .scale(scale, scale)
                    .build());
        };
    }
    
    /**
     * Parses a length as a number of points. Note that this does not handle
     * relative units such as "auto", "%", "em", "ex" properly.
     * @param len input length
     * @return equivalent # of points
     */
    public static double parseLength(String len) {
        if (len == null) {
            LOG.log(Level.WARNING, "Null length: {0}", len);
            return 0;
        }
        String uselen = len.trim();
        Matcher m = LEN_PATTERN.matcher(uselen);
        if (m.matches()) {
            if ("auto".equals(len) || len.contains("%")) {
                LOG.log(Level.WARNING, "Unsupported length: {0}", len);
                return 0;
            }
            String sunit = m.group(2);
            Unit unit = unit(sunit);
            String num = sunit == null ? len : len.substring(0, len.length()-sunit.length());
            Double val = Double.parseDouble(num);
            return unit.toPixels(val);
        }
        return 0;
    }
    
    private static Unit unit(String unit) {
        if (unit == null) {
            return Unit.PT;
        }
        for (Unit u : Unit.values()) {
            if (u.name.equalsIgnoreCase(unit)) {
                return u;
            }
        }
        LOG.log(Level.WARNING, "Invalid unit: {0}", unit);
        return Unit.PT;
    }
    
    private static enum Unit {
        PX("px", 1.3f),
        EM("em", 11.955f),
        EX("ex", 11.955f),
        PCT("%", 1f),
        IN("in", 72f),
        CM("cm", 28.3464f),
        MM("mm", 2.83464f),
        PT("pt", 1f),
        PC("pc", 12f);
        
        private String name;
        private float scale;

        private Unit(String name, float scale) {
            this.name = name;
            this.scale = scale;
        }
        
        private double toPixels(double val) {
            return val*scale;
        }
    }
    
}

File: blaise-svg-kt/src/main/java/com/googlecode/blaisemath/graphics/svg/SvgUtils.kt
/**
 * SvgUtils.java
 * Created on May 5, 2015
 */
package com.googlecode.blaisemath.graphics.svg
/*
 * #%L
 * BlaiseSvg
 * --
 * Copyright (C) 2014 - 2021 Elisha Peterson
 * --
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
import com.google.common.graph.ElementOrder
import com.google.common.graph.EndpointPair
import com.google.common.graph.GraphBuilder
import com.google.common.graph.Graphs
import com.google.common.graph.ImmutableGraph
import com.google.common.graph.MutableGraph
import com.googlecode.blaisemath.app.ApplicationMenuConfig
import com.googlecode.blaisemath.graph.ContractedGraphTest
import com.googlecode.blaisemath.graph.GraphComponents
import com.googlecode.blaisemath.graph.GraphGenerator
import com.googlecode.blaisemath.graph.GraphMetric
import com.googlecode.blaisemath.graph.GraphMetrics
import com.googlecode.blaisemath.graph.GraphNodeMetric
import com.googlecode.blaisemath.graph.GraphNodeStats
import com.googlecode.blaisemath.graph.GraphServices
import com.googlecode.blaisemath.graph.GraphSubsetMetric
import com.googlecode.blaisemath.graph.GraphUtilsTest
import com.googlecode.blaisemath.graph.IterativeGraphLayout
import com.googlecode.blaisemath.graph.IterativeGraphLayoutState
import com.googlecode.blaisemath.graph.NodeInGraph
import com.googlecode.blaisemath.graph.OptimizedGraph
import com.googlecode.blaisemath.graph.StaticGraphLayout
import com.googlecode.blaisemath.graph.SubgraphTest
import com.googlecode.blaisemath.graph.app.AnimationUtils
import com.googlecode.blaisemath.graph.app.GraphApp
import com.googlecode.blaisemath.graph.app.GraphAppCanvas
import com.googlecode.blaisemath.graph.app.GraphAppFrameView
import com.googlecode.blaisemath.graph.app.MetricScaler
import com.googlecode.blaisemath.graph.generate.AbstractGraphGenerator
import com.googlecode.blaisemath.graph.generate.CompleteGraphGenerator
import com.googlecode.blaisemath.graph.generate.CycleGraphGenerator
import com.googlecode.blaisemath.graph.generate.DefaultGeneratorParameters
import com.googlecode.blaisemath.graph.generate.DegreeDistributionGenerator
import com.googlecode.blaisemath.graph.generate.DegreeDistributionGenerator.DegreeDistributionParameters
import com.googlecode.blaisemath.graph.generate.EdgeCountGenerator
import com.googlecode.blaisemath.graph.generate.EdgeLikelihoodGenerator
import com.googlecode.blaisemath.graph.generate.EdgeLikelihoodGenerator.EdgeLikelihoodParameters
import com.googlecode.blaisemath.graph.generate.EmptyGraphGenerator
import com.googlecode.blaisemath.graph.generate.ExtendedGeneratorParameters
import com.googlecode.blaisemath.graph.generate.GraphGenerators
import com.googlecode.blaisemath.graph.generate.GraphGrowthRule
import com.googlecode.blaisemath.graph.generate.GraphSeedRule
import com.googlecode.blaisemath.graph.generate.HopGrowthRule
import com.googlecode.blaisemath.graph.generate.PreferentialAttachmentGenerator
import com.googlecode.blaisemath.graph.generate.PreferentialAttachmentGenerator.PreferentialAttachmentParameters
import com.googlecode.blaisemath.graph.generate.ProximityGraphGenerator.ProximityGraphParameters
import com.googlecode.blaisemath.graph.generate.StarGraphGenerator
import com.googlecode.blaisemath.graph.generate.WattsStrogatzGenerator
import com.googlecode.blaisemath.graph.generate.WattsStrogatzGenerator.WattsStrogatzParameters
import com.googlecode.blaisemath.graph.generate.WheelGraphGenerator
import com.googlecode.blaisemath.graph.layout.CircleLayout
import com.googlecode.blaisemath.graph.layout.CircleLayout.CircleLayoutParameters
import com.googlecode.blaisemath.graph.layout.GraphLayoutConstraints
import com.googlecode.blaisemath.graph.layout.GraphLayoutManager
import com.googlecode.blaisemath.graph.layout.IterativeGraphLayoutManager
import com.googlecode.blaisemath.graph.layout.IterativeGraphLayoutService
import com.googlecode.blaisemath.graph.layout.LayoutRegion
import com.googlecode.blaisemath.graph.layout.PositionalAddingLayout
import com.googlecode.blaisemath.graph.layout.RandomBoxLayout
import com.googlecode.blaisemath.graph.layout.RandomBoxLayout.BoxLayoutParameters
import com.googlecode.blaisemath.graph.layout.SpringLayoutParameters
import com.googlecode.blaisemath.graph.layout.SpringLayoutPerformanceTest
import com.googlecode.blaisemath.graph.layout.SpringLayoutState
import com.googlecode.blaisemath.graph.layout.StaticSpringLayout
import com.googlecode.blaisemath.graph.layout.StaticSpringLayout.StaticSpringLayoutParameters
import com.googlecode.blaisemath.graph.metrics.AbstractGraphMetric
import com.googlecode.blaisemath.graph.metrics.AbstractGraphNodeMetric
import com.googlecode.blaisemath.graph.metrics.AdditiveSubsetMetricTest
import com.googlecode.blaisemath.graph.metrics.BetweenCentrality
import com.googlecode.blaisemath.graph.metrics.BetweenCentralityTest
import com.googlecode.blaisemath.graph.metrics.ClosenessCentrality
import com.googlecode.blaisemath.graph.metrics.ClosenessCentralityTest
import com.googlecode.blaisemath.graph.metrics.ClusteringCoefficient
import com.googlecode.blaisemath.graph.metrics.CooperationMetric
import com.googlecode.blaisemath.graph.metrics.DecayCentrality
import com.googlecode.blaisemath.graph.metrics.EigenCentrality
import com.googlecode.blaisemath.graph.metrics.EigenCentralityTest
import com.googlecode.blaisemath.graph.metrics.GraphCentrality
import com.googlecode.blaisemath.graph.metrics.GraphCentralityTest
import com.googlecode.blaisemath.graph.metrics.GraphDiameter
import com.googlecode.blaisemath.graph.metrics.SubsetMetrics
import com.googlecode.blaisemath.graph.metrics.SubsetMetrics.AdditiveSubsetMetric
import com.googlecode.blaisemath.graph.metrics.SubsetMetrics.ContractiveSubsetMetric
import com.googlecode.blaisemath.graph.metrics.SubsetMetricsTest
import com.googlecode.blaisemath.graph.test.DynamicGraphTestFrame
import com.googlecode.blaisemath.graph.test.GraphTestFrame
import com.googlecode.blaisemath.graph.test.MyTestGraph
import com.googlecode.blaisemath.graph.util.Matrices
import com.googlecode.blaisemath.graph.view.GraphComponent
import com.googlecode.blaisemath.graph.view.VisualGraph
import com.googlecode.blaisemath.graph.view.WeightedEdgeStyler
import com.googlecode.blaisemath.graphics.editor.AttributeSetPropertyModelTestFrame
import com.googlecode.blaisemath.graphics.swing.AnchorTestFrame
import com.googlecode.blaisemath.graphics.testui.ContextMenuTestFrame
import com.googlecode.blaisemath.graphics.testui.HelloWorldTest
import com.googlecode.blaisemath.graphics.testui.PanAndZoomTestFrame
import com.googlecode.blaisemath.graphics.testui.SelectionTestFrame
import com.googlecode.blaisemath.graphics.testui.TooltipTestFrame
import com.googlecode.blaisemath.style.Marker
import com.googlecode.blaisemath.svg.HelloWorldSvg
import com.googlecode.blaisemath.svg.SvgPath
import com.googlecode.blaisemath.svg.SvgPathTest
import com.googlecode.blaisemath.svg.SvgRootTest
import com.googlecode.blaisemath.svg.SvgTool
import com.googlecode.blaisemath.test.AssertUtils
import com.googlecode.blaisemath.ui.PropertyActionPanel
import com.googlecode.blaisemath.util.geom.AffineTransformBuilder
import junit.framework.TestCase
import org.apache.commons.math.stat.descriptive.SummaryStatistics
import org.junit.BeforeClass
import java.awt.geom.Path2D
import java.awt.geom.Point2D
import java.util.logging.Level
import java.util.logging.Logger
import java.util.regex.Pattern

/**
 * Utility class for Svg paths.
 * @author petereb1
 */
object SvgUtils {
    private val LOG = Logger.getLogger(SvgUtils::class.java.name)
    private val LEN_PATTERN = Pattern.compile("^auto$|^[+-]?[0-9]+\\.?([0-9]+)?(px|em|ex|%|in|cm|mm|pt|pc)?$")

    /**
     * Convert the given path string to a marker with given size.
     * @param svgPath Svg path string
     * @param sz size of the resulting marker, as the side length/diameter
     * @return marker
     */
    fun pathToMarker(svgPath: String?, sz: Float): Marker? {
        val path: Path2D = SvgPath.Companion.shapeConverter().convert(SvgPath(svgPath))
        if (path == null) {
        }
        return Marker { point: Point2D?, orientation: Double, markerRadius: Float ->
            val scale = markerRadius / (.5 * sz)
            path.createTransformedShape(AffineTransformBuilder()
                    .translate(point.getX() - markerRadius, point.getY() - markerRadius)
                    .scale(scale, scale)
                    .build())
        }
    }

    /**
     * Parses a length as a number of points. Note that this does not handle
     * relative units such as "auto", "%", "em", "ex" properly.
     * @param len input length
     * @return equivalent # of points
     */
    fun parseLength(len: String?): Double {
        if (len == null) {
            LOG.log(Level.WARNING, "Null length: {0}", len)
            return 0
        }
        val uselen = len.trim { it <= ' ' }
        val m = LEN_PATTERN.matcher(uselen)
        if (m.matches()) {
            if ("auto" == len || len.contains("%")) {
                LOG.log(Level.WARNING, "Unsupported length: {0}", len)
                return 0
            }
            val sunit = m.group(2)
            val unit = unit(sunit)
            val num = if (sunit == null) len else len.substring(0, len.length - sunit.length)
            val `val` = num.toDouble()
            return unit.toPixels(`val`)
        }
        return 0
    }

    private fun unit(unit: String?): Unit? {
        if (unit == null) {
            return Unit.PT
        }
        for (u in Unit.values()) {
            if (u.name.equals(unit, ignoreCase = true)) {
                return u
            }
        }
        LOG.log(Level.WARNING, "Invalid unit: {0}", unit)
        return Unit.PT
    }

    private enum class Unit(private val name: String?, private val scale: Float) {
        PX("px", 1.3f), EM("em", 11.955f), EX("ex", 11.955f), PCT("%", 1f), IN("in", 72f), CM("cm", 28.3464f), MM("mm", 2.83464f), PT("pt", 1f), PC("pc", 12f);

        private fun toPixels(`val`: Double): Double {
            return `val` * scale
        }
    }
}